### PR TITLE
Spark: Remove ImmutableMap from SparkPartition for Kryo SerDe

### PR DESCRIPTION
**Pull request by kbendick**
This PR replaces https://github.com/apache/iceberg/pull/3597, as @dubeme has not responded about some minor changes and we need this patch prior to the upcoming 0.13.0 release.

I have added them as a co-author on this PR so that they still get credit for their contribution 😄 

Issue #3586 shows situation where Kyro is unable to serialize the SparkPartition object (@rdblue points it out from the stack trace). This commit replaces the ImmutableMap with HashMap and adds a test for round trip Kryo serialization and Java serialization for `SparkPartition` objects.

I also tested this by running the `add_files` test suite with Kryo serialization enabled. The tests failed until this patch was applied. Unfortunately, I couldn't add a test there as the usual `withSQLConf` method did not pick up the change in `spark.serializer`.

Co-authored-by: Dubem Enyekwe <dubem@dubemenyekwe.com>

cc @rdblue @jackye1995 @dubeme

### CODE REVIEW

**Comment by kbendick** [comment]
Hmmm it seems like when I rebased, it might have removed the co-authorship from @dubeme.

Before we merge this in, can we please ensure that `Co-authored-by: Dubem Enyekwe <dubem@dubemenyekwe.com>` is included so they get credit?

I can update it on my end somehow if need be.

File: spark/v3.2/spark/src/test/java/org/apache/iceberg/spark/TestSparkPartitionSerialization.java
Comment (jackye1995) [inline spark/v3.2/spark/src/test/java/org/apache/iceberg/spark/TestSparkPartitionSerialization.java:31]: I think we can just use `TestSparkTableUtil` as test class name

**Review by jackye1995 — COMMENTED** [review]
Given this is a one-line change, could you also fix the other version?

File: spark/v3.2/spark/src/test/java/org/apache/iceberg/spark/TestSparkPartitionSerialization.java
Comment (kbendick) [reply spark/v3.2/spark/src/test/java/org/apache/iceberg/spark/TestSparkPartitionSerialization.java:31]: Yeah that seems reasonable.

**Comment by kbendick** [comment]
> Given this is a one-line change, could you also fix the other versions?

Sure thing. I almost did but I wasn’t sure where exactly we’re drawing the line. But this is small and if people need it, they need it for all versions (or can edit one or more out).

Will update.

File: spark/v3.2/spark/src/test/java/org/apache/iceberg/spark/TestSparkPartitionSerialization.java
Comment (kbendick) [reply spark/v3.2/spark/src/test/java/org/apache/iceberg/spark/TestSparkPartitionSerialization.java]: Updated it to be `TestSparkTableUtil`. In Spark 2.4, there already is a `TestSparkTableUtil` so I just added it into that class (those tests are somewhere more specific in Spark 3+).

**Review by jackye1995 — APPROVED** [review]
looks good to me, will merge on Monday if no one objects.

**Review by rdblue — APPROVED** [review]

**Comment by rdblue** [comment]
Thanks for getting this finished, @kbendick!

**Comment by dubeme** [comment]
Thanks @kbendick  for stepping in in my absence

**Pull request merged**
